post: rust-libp2p webrtc in the browser (#117)
libp2p/blog: src/_blog/rust-libp2p-browser-to-server.md

---
tags:
  - rust-libp2p
  - rust
  - libp2p
  - browser
  - webrtc
  - webassembly
  - wasm

> title: Rust-libp2p in the Browser with WebRTC!

Reviewer: ```suggestion
title: rust-libp2p in the Browser with WebRTC!
```

I don't think we capitalize the _r_ anywhere.

> description: News about how rust-libp2p now connects with the network via WebRTC

Reviewer: ```suggestion
description: News about how rust-libp2p in the browser connects to nodes in a network using WebRTC
```

I suggest:

- Not talking about _the_ network. I.e. there are many.
- Emphasizing that rust-libp2p **in the browser** is the special part here.

Don't feel strongly about the exact wording.

date: 2023-11-07
permalink: ''
translationKey: ''
header_image: /rust-libp2p.jpeg
author: DougAnderson444
---

# Rust-libp2p WebRTC now available in the Browser

> WebAssembly WebRTC is now available for use as a browser transport with rust-libp2p! Before now, WebRTC was only available on the server in `rust-libp2p`, but after months of coding and reviews, we are pleased to annouce it is available on both the server and in the browser.

Reviewer: ```suggestion
rust-libp2p running in the browser can now establish WebRTC connections! Before now, WebRTC was only available on the server in `rust-libp2p`, but after months of coding and reviews, we are pleased to annouce it is available on both the server and in the browser.
```

Would be easier for me to understand. What do you think?

Author: I like it. It's a lot of terms to squeeze into one breath, I think your rendition does the trick very nicely

## Why WebRTC for libp2p?

### Data Channels

WebRTC's main purpose in peer-to-peer communications is for connecting media (audio and video). However, there is also a powerful [data channel](https://webrtc.org/getting-started/data-channels) API available which allows users to exchange data streams instead of audio/video. This is what libp2p uses at the transport layer. By tapping into this data channel API, we have been able to build libp2p Transports over WebRTC.

### Server WebRTC vs Browser WebRTC

Using WebRTC on the server involves using a Rust crate (currently [webrtc-rs](https://github.com/webrtc-rs/webrtc)), but in the browser it is available to Rust through browser bindings from a crate called [`web-sys`](https://docs.rs/web-sys/latest/web_sys/). By tying the libp2p code to these WebRTC bindings, we have been able to connect in the browser over WebRTC in pure Rust compiled to WebAssembly!

### Benefits of WebRTC over websockets and webtransport

> One of the main benefits of using WebRTC on the server is that the it does not need TLS certificates from a DNS domain -- you can run a home server over WebRTC without a dotcom! Instead, with `rust-libp2p` we can use self-signed certificates to provide the secure connections. Websockets however would require TLS certificates.

Reviewer: ```suggestion
One of the main benefits of using WebRTC on the server is that it does not need TLS certificates from a DNS domain -- you can run a home server over WebRTC without a dotcom! Instead, with `rust-libp2p` we can use self-signed certificates to provide the secure connections. Websockets however would require TLS certificates.
```

> One of the main benefits of using WebRTC on the server is that the it does not need TLS certificates from a DNS domain -- you can run a home server over WebRTC without a dotcom! Instead, with `rust-libp2p` we can use self-signed certificates to provide the secure connections. Websockets however would require TLS certificates.

Reviewer: Note that authorities provide TLS certificates for IPs, not just domains. Overall explanation seems clunky. I like the "self-signed" wording. Maybe you can copy parts of https://blog.libp2p.io/libp2p-webrtc-browser-to-server/#self-signed-certificate?

Author: Thanks for these, I will re-word it to make it less clunky. I am trying to convey that there should be fewer steps and resources to set up a WebRTC node, say, at home, compared to a Websocket secure node.

> The second major benefot of WebRTC is it enables (eventual\*) direct browser-to-browser communication. Though webtransport can connect with self-signed certificates, it cannot connect browser-to-browser like WebRTC can.

Reviewer: ```suggestion
The second major benefit of WebRTC is it enables (eventual\*) direct browser-to-browser communication. Though webtransport can connect with self-signed certificates, it cannot connect browser-to-browser like WebRTC can.
```

> The second major benefot of WebRTC is it enables (eventual\*) direct browser-to-browser communication. Though webtransport can connect with self-signed certificates, it cannot connect browser-to-browser like WebRTC can.

Reviewer: This is the first mention of WebTransport. Did you meant WebSocket?

Author: I did mention WebTransport in a heading, but I can re-iterate the comparison again in the para intro

> The second major benefot of WebRTC is it enables (eventual\*) direct browser-to-browser communication. Though webtransport can connect with self-signed certificates, it cannot connect browser-to-browser like WebRTC can.

Reviewer: I suggest focusing this blog post on what exists, i.e. `/webrtc-direct` and not `/webrtc`.

Author: But then I loose one of my main benefits 🫤

In all seriousness, I get what you are saying, but I also wonder if people understand that WebRTC can do this and that it is a main benefit (albeit a forthcoming benefit....). If we don't mention it, readers may wonder what the point is over Websockets & Webtransport.

I could perhaps just mention it here and then focus the rest of the `/webrtc-direct` parts?

> \*Note: This is a common point of confusion for WebRTC when connecting two browsers together. Although the browser can _eventually_ connect directly, to set up this direct connection we need a signalling server. In the case of libp2p, we can use a libp2p server to conduct this signalling by using circuit relay. But we always need the WebRTC server first. This server can be any libp2p node that is not running in the browser, it could be a home computer for example.

> With WebRTC rust-libp2p now available in the browser, we can enable developers to build full stack applications in rust without the need for a domain name or external server.

Reviewer: ```suggestion
With WebRTC rust-libp2p now available in the browser, we can enable developers to build full stack applications in rust without the need for certificate authorities or external server.
```

### Benefits of Full Stack rust-webrtc

Before now, the only way the browser could connect to a rust-libp2p server over WebRTC was to use [js-libp2p](https://github.com/libp2p/js-libp2p).

> However, full stack rust-libp2p WebRTC is so beneficial. First, troubleshooting was much more difficult to do in two stacks, two languages, two styles of implementation. Even though js-libp2p and rust-libp2p implement the same libp2p spec, it's much more difficult to switch between languages when troubleshooting bugs and data flow. With full-stack rust-libp2p, we can now build applications that run in the browser and on the server in rust, using the same language and style of implementation on both the client and the server. This makes both developing applications fatser and easier as we can re-use code in both the browser and on the server. In fact, we were able to move a lot of code out of the server crate into a [common WebRTC utilities crate within the rust-libp2p repo](https://github.com/libp2p/rust-libp2p/tree/master/misc/webrtc-utils), This will apply to libp2p applications as well!

Reviewer: ```suggestion
However, full stack rust-libp2p WebRTC is so beneficial. First, troubleshooting was much more difficult to do in two stacks, two languages, two styles of implementation. Even though js-libp2p and rust-libp2p implement the same libp2p spec, it's much more difficult to switch between languages when troubleshooting bugs and data flow. With full-stack rust-libp2p, we can now build applications that run in the browser and on the server in rust, using the same language and style of implementation on both the client and the server. This makes both developing applications faster and easier as we can re-use code in both the browser and on the server. In fact, we were able to move a lot of code out of the server crate into a [common WebRTC utilities crate within the rust-libp2p repo](https://github.com/libp2p/rust-libp2p/tree/master/misc/webrtc-utils), This will apply to libp2p applications as well!
```

> However, full stack rust-libp2p WebRTC is so beneficial. First, troubleshooting was much more difficult to do in two stacks, two languages, two styles of implementation. Even though js-libp2p and rust-libp2p implement the same libp2p spec, it's much more difficult to switch between languages when troubleshooting bugs and data flow. With full-stack rust-libp2p, we can now build applications that run in the browser and on the server in rust, using the same language and style of implementation on both the client and the server. This makes both developing applications fatser and easier as we can re-use code in both the browser and on the server. In fact, we were able to move a lot of code out of the server crate into a [common WebRTC utilities crate within the rust-libp2p repo](https://github.com/libp2p/rust-libp2p/tree/master/misc/webrtc-utils), This will apply to libp2p applications as well!

Reviewer: I would expect the first benefit to be, that one can share code? I.e. write once run (/debug :P) anywhere?

Author: I've re-worded this whole paragraph to streamline that point, good call

> ## How does libp2p use WebRTC?
> 
> The bindings to [web-sys](https://docs.rs/web-sys/latest/web_sys/) are implemented as a transport, which means that you can connect to the libp2p network from the browser as you would any other transport in `rust-libp2p`. There is a full stack example [in the repo](https://github.com/libp2p/rust-libp2p/tree/master/examples/browser-webrtc).

Reviewer: This paragraph is confusing to me. I think it is worth reintroducing web-sys here. How about something along the lines of:

```suggestion
## How does rust-libp2p in the browser use WebRTC?

- ~there is web-sys
- enables one to bind to browser apis
- rust-libp2p wraps web-sys, implementing the `Transport` trade
- Thus allows any application protocol (e.g. Kademlia) to make use of webrtc connections.

There is a full stack example [in the repo](https://github.com/libp2p/rust-libp2p/tree/master/examples/browser-webrtc).
```

## Tutorial

Code walk through?

## What's next?

> There is still work to be done to connect browser-to-browser in rust using `web-sys`. The [spec](https://github.com/libp2p/specs/tree/master/webrtc) is complete, and just needs to be implemented! We are always looking for more help from other community members to make this happen.

Reviewer: How about linking to https://github.com/libp2p/rust-libp2p/issues/4389?

Author: Yes!!

---
tags:
  - rust-libp2p
  - rust
  - libp2p
  - browser
  - webrtc
  - webassembly
  - wasm
title: rust-libp2p in the Browser with WebRTC!
description: News about how rust-libp2p in the browser connects to nodes in a network using WebRTC
date: 2023-12-07
permalink: ''
translationKey: ''
header_image: /rust-libp2p.jpeg
author: DougAnderson444
---

# Rust-libp2p WebRTC now available in the Browser

> We are excited to announce that [rust-libp2p](https://github.com/libp2p/rust-libp2p) running in the browser can now establish [WebRTC connections](https://webrtc.org/). Before now, WebRTC was only available on the server in [rust-libp2p](https://github.com/libp2p/rust-libp2p), but after months of coding and reviews, we are pleased to annouce it is available on both the server and in the browser from rust-libp2p.

Reviewer 2: ```suggestion
We are excited to announce that [rust-libp2p](https://github.com/libp2p/rust-libp2p) running in the browser can now establish [WebRTC connections](https://webrtc.org/). Before now, WebRTC was only available on the server in [rust-libp2p](https://github.com/libp2p/rust-libp2p), but after months of coding and reviews, we are pleased to announce it is available on both the server and in the browser from rust-libp2p.
```

# Why WebRTC for rust-libp2p?

> There are a few different Transports to choose from, each with their pros and cons. So, why WebRTC? Let's dive into some of the benefits of WebRTC and why we decided it was important to implement as a Transport within rust-libp2p for use in the browser.

Reviewer 3: Is the Letter Case of `Transport` on purpose? Reads a bit odd but I am not a native speaker :D

Reviewer 2: :+1: if it's meant to refer to an API term, maybe something like:

> There are a few different options for a `Transport` you can choose from,

(note how it is put in a code block)

> There are a few different Transports to choose from, each with their pros and cons. So, why WebRTC? Let's dive into some of the benefits of WebRTC and why we decided it was important to implement as a Transport within rust-libp2p for use in the browser.

Reviewer 2: ```suggestion
There are a few different options for a `Transport` to choose from, each with its own pros and cons. So, why WebRTC? Let's dive into some of the benefits of WebRTC and why we decided it was important to implement it as a `Transport` within rust-libp2p for use in the browser.
```

Author: Good idea, I will use the code block with the `T`

## Benefits of WebRTC over WebSockets and WebTransport

There are certain benefits of using WebRTC instead WebSockets and WebTransport.

> The benefit of using WebRTC instead of _WebSockets_ on the _server_ is that WebRTC uses self-signed TLS certificates, eliminating the need for users to set up additional services like DNS and Let's Encrypt, which we would have to do with WebSockets to make it available via `wss://`. This is great because it means users can run a public rust-libp2p WebRTC server out of the box and have Transport access to that server from outside their network without additional steps or resources.

Reviewer 3: The start of this sentence is a bit repetitive with the use of "benefit".

I'd suggest to highlight both benefits explained in this paragraph with a few keywords and then explain them in more detail, something like (feel free to expand, I just focused on what I think are the essential bits of information):

1. Easier node operation: Modern browsers require the use of TLS in many cases. In the case of WebSockets, the certificates need to be signed by a CA that is in the browsers trust-chain which typically means you'll need to set up DNS and Let's Encrypt (or similar). With WebRTC, we can use self-signed certificates, removing the need for DNS and trusted CAs! WebTransport also allows the use of self-signed certificates but `rust-libp2p` does not (yet) support WebTransport as a server-side transport. Subscribe to https://github.com/libp2p/rust-libp2p/issues/2993 for updates.
2. Improved connectivity: WebRTC allows us to connect two browsers directly. Typically, connecting two nodes requires one of them to "listen" on one or more interfaces and sharing its listen address with the other node. Browsers however don't provide direct access to a machine's network interfaces. With WebRTC, browsers expose an API to establish a direct connection to another browser with the help of a signalling server.

Author: I like it! I'll re-word this a bit to focus on these two areas
